[processor/attributes] Support attributes set by server authenticators

svrakitin opened the pull request:
**Description:**

Support looking up attributes from `client.Info.Auth` which is set by extensions implementing `configauth.ServerAuthenticator`. This way we can attach authentication of the exporter (e.g. OIDC subject) to the telemetry.

User can refer to these attributes by prefixing keys with `auth.` when using `from_context`:

```yaml
attributes/subject:
  actions:
  - key: enduser.id
    from_context: auth.subject
    action: insert
```

For consistency I added `metadata.` prefix to refer to actual metadata/header keys. This prefix is optional for backwards compatibility.

**Testing:**

I added tests to `attraction` package which has actual extraction logic.

Closes #9241

svrakitin commented on internal/coreinternal/attraction/attraction.go:
```diff
@@ -311,8 +311,38 @@ func (ap *AttrProc) Process(ctx context.Context, logger *zap.Logger, attrs pcomm
+		case []string:
+			vals = a
+		default:
+			return pcommon.Value{}, false
```
What to do with non-string typed attributes? At the moment we don't have any use cases as auth extensions either return `string` or `[]string` attributes.

Should I match on raw type and produce corresponding value? I would probably do that if `newValueFromRaw` was public. https://github.com/open-telemetry/opentelemetry-collector/blob/6fed15bf12d807da8ac834fd776eb1423187ca19/pdata/internal/common.go#L135

pmm-sumo replied:
I wonder if this might ever happen. Perhaps we should just log that unexpected type was observed (ideally only once for a while, e.g. using logging sampler or other means)

jpkrohling replied:
I consider this an error condition and users should be appropriately warned if this ever happens.

svrakitin replied:
Added log warning.

jpkrohling commented on processor/attributesprocessor/README.md:
```diff
@@ -46,8 +46,10 @@ For the actions `insert`, `update` and `upsert`,
   # FromContext specifies the context value to use to populate the attribute value. 
-  # The context values would be searched in receiver's transport protocol additional information
-  # like GRPC Metadata or HTTP Headers. 
+  # If key is prefixed with `metadata.` or has no prefix then values are searched 
```
```suggestion
  # If the key is prefixed with `metadata.`, the values are searched 
```

If we ever intend to remove the support for unqualified keys, it's better to not mention the "no prefix" here.

jpkrohling commented on processor/attributesprocessor/README.md:
```diff
@@ -46,8 +46,10 @@ For the actions `insert`, `update` and `upsert`,
-  # The context values would be searched in receiver's transport protocol additional information
-  # like GRPC Metadata or HTTP Headers. 
+  # If key is prefixed with `metadata.` or has no prefix then values are searched 
+  # in receiver's transport protocol additional information like GRPC Metadata or HTTP Headers. 
```
```suggestion
  # in the receiver's transport protocol additional information like gRPC Metadata or HTTP Headers. 
```

jpkrohling commented on processor/attributesprocessor/README.md:
```diff
@@ -46,8 +46,10 @@ For the actions `insert`, `update` and `upsert`,
-  # like GRPC Metadata or HTTP Headers. 
+  # If key is prefixed with `metadata.` or has no prefix then values are searched 
+  # in receiver's transport protocol additional information like GRPC Metadata or HTTP Headers. 
+  # If key is prefixed with `auth.` then values are searched
```
```suggestion
  # If the key is prefixed with `auth.`, the values are searched
```

jpkrohling commented on processor/attributesprocessor/README.md:
```diff
@@ -46,8 +46,10 @@ For the actions `insert`, `update` and `upsert`,
+  # If key is prefixed with `metadata.` or has no prefix then values are searched 
+  # in receiver's transport protocol additional information like GRPC Metadata or HTTP Headers. 
+  # If key is prefixed with `auth.` then values are searched
+  # in authentication information set by receiver authenticator.
```
```suggestion
  # in the authentication information set by the server authenticator. Refer to the server authenticator's documentation part of your pipeline for more information about which attributes are available.
```

jpkrohling commented on processor/attributesprocessor/testdata/config.yaml:
```diff
@@ -342,6 +342,15 @@ processors:
+      from_context: metadata.origin
+      action: insert
+    - key: enduser.id
+      from_context: auth.subject
```
Can you refer to a specific server authenticator that produces `subject`?

svrakitin replied:
I hope this is useful enough. Probably needs help with wording.

jpkrohling reviewed (commented):
LGTM, just the documentation could be a bit better. Thanks for this PR!

svrakitin commented:
@jpkrohling I added logger to the `AttrProc` and removed it from the `Process` method. Now logger is injected from the processor.

Thanks for suggestions to the docs.

pmm-sumo commented on internal/coreinternal/attraction/attraction.go:
```diff
@@ -305,14 +312,49 @@ func (ap *AttrProc) Process(ctx context.Context, logger *zap.Logger, attrs pcomm
+		case []string:
+			vals = a
+		default:
+			ap.logger.Warn(
```
This could potentially generate a lot of log lines. We've previously recommended using (to the extent of copy/pasting) [sampled logger](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry.go#L61-L78) though with so many cases it's relevant perhaps we just put it into some common helper package

svrakitin replied:
Thanks for the tip, I missed that previously.

jpkrohling commented on internal/coreinternal/attraction/attraction_test.go:
```diff
@@ -86,7 +86,7 @@ func TestAttributes_InsertValue(t *testing.T) {
 	}
 
-	ap, err := NewAttrProc(cfg)
+	ap, err := NewAttrProc(cfg, nil)
```
I won't block the PR for this unless @pmm-sumo agrees with me, but I would prefer the test to be passing zap.NewNop() instead of nil. Remember that your test code shows the good practice for consumers of the main code, and by passing nil, you are sending the message that nil is OK, when it really isn't.

svrakitin replied:
Sure, I don't have a preference here. Gonna update.

pmm-sumo commented on internal/coreinternal/attraction/attraction.go:
```diff
@@ -168,13 +170,21 @@ type attributeAction struct {
-func NewAttrProc(settings *Settings) (*AttrProc, error) {
-	var attributeActions []attributeAction
+func NewAttrProc(settings *Settings, logger *zap.Logger) (*AttrProc, error) {
+	if logger == nil {
```
I think now we might assume that logger is never nil?

jpkrohling replied:
It's good to check for nil, but not sure about the outcome: you do want to have access to a logger, it's not an optional parameter, so, you should return an error if it's nil.

bogdandrutu replied:
I agree, this seems to be unrelated, consider to split a separate PR.

rkukura reviewed (approved):
This all looks good to me - very similar to what I planned to implement when filing https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9241, except that I didn't anticipate the logging changes. Please let me know if I can help get this merged.

jpkrohling commented on internal/coreinternal/attraction/attraction.go:
```diff
@@ -305,14 +332,50 @@ func (ap *AttrProc) Process(ctx context.Context, logger *zap.Logger, attrs pcomm
+				zap.String("name", attrName),
+				zap.String("type", fmt.Sprintf("%T", a)),
+			)
+			ap.logger.WithOptions()
```
What's this line doing?

jpkrohling commented:
PR rebased

svrakitin commented:
@jpkrohling What is left to get it merged?

pmm-sumo commented:
Juraci is now on parental leave :) But since we have 3 approvals (including two from approvers/contributors) and passing tests, I'm marking this as `ready-to-merge`

bogdandrutu commented on internal/coreinternal/attraction/attraction.go:
```diff
@@ -267,35 +277,52 @@ func NewAttrProc(settings *Settings) (*AttrProc, error) {
+	return ap, nil
+}
+
+func loggerWithSampling(logger *zap.Logger) *zap.Logger {
```
Why this random configuration?

bogdandrutu commented on internal/coreinternal/attraction/attraction.go:
```diff
@@ -267,35 +277,52 @@ func NewAttrProc(settings *Settings) (*AttrProc, error) {
 		switch action.Action {
 		case DELETE:
-			attrs.Delete(action.Key)
+			attrs.Remove(action.Key)
```
This was fixed in a separate PR, and also unrelated to your main PR.

bogdandrutu commented:
@pmm-sumo there are lots of unrelated changes, and some random configurations for logger (why do we need sampling here and not in other places? add debug logs if too many, record metrics, etc.)

svrakitin commented:
@bogdandrutu There was a suggestion to log invalid atrribute types if they are returned by ServerAuthenticator: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9420#discussion_r858965522

I don’t care much about sampling and can drop it. It was taken from the suggestion in one of the comments above.

I will reduce this PR to changes I originally introduced and leave further changes to maintainers who have a better idea than me.

bogdandrutu commented:
@svrakitin it is not about "better idea than me", it is about keeping the PRs focus on what we want to implement :)

svrakitin committed:
[processor/attributes] Support attributes set by server authenticators

pmm-sumo commented:
Good point @bogdandrutu, apologies for messing up with tagging and thanks for bringing this up. 

On the sampled logger thing, I believe it [stems from this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9420/files#r859067416). I mentioned sampled logging since this was something brought to my attention in the past (don't recall which PR but I can look it up for context).

In principle, I think that user should know that something is wrong and because of that `debug` logs are not a good fit.  I also heard some comments that the collector is sometimes too quiet over what happens in the components and some users anticipate more logging. At the same time, we want to limit the impact of that log if happens frequently, hence (aggressive) sampled logger and `warn` level. 

So we have two paths here:
* always use `debug` level for such kind of logs
* use sampled logger and move it to a common package, so other components could use it easily for potentially noisy logs

I was inclined towards the latter (in a separate PR). There are some components doing that already and I was aiming at doing the cleanup after this get merged. Before I open the issue wanted to check what is your preference @bogdandrutu

bogdandrutu reviewed (commented):
This PR looks much better now, @pmm-sumo please re-approve since you are familiar, and add the label again

svrakitin committed:
Merge branch 'main' into attributesprocessor-auth-from-context

bogdandrutu committed:
Merge branch 'main' into attributesprocessor-auth-from-context

Commit:
Merge branch 'main' into attributesprocessor-auth-from-context
